cmd/auth-provider-gcp/plugin/plugin_test.go

```diff
@@ -0,0 +1,213 @@
+	return auth.Username == expectedUsername
+}
+
+func passwordMatches(expectedPassword string, auth credentialproviderapi.AuthConfig) bool {
+	return auth.Password == expectedPassword
+}
+
+func TestContainerRegistry(t *testing.T) {
+	// Taken from from pkg/credentialprovider/gcp/metadata_test.go in k/k
+	registryURL := "container.cloud.google.com"
+	token := &gcpcredential.TokenBlob{AccessToken: dummyToken} // Fake value for testing.
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Only serve the URL key and the value endpoint
+		if scopeEndpoint == r.URL.Path {
+			w.WriteHeader(http.StatusOK)
+			w.Header().Set("Content-Type", "application/json")
+			fmt.Fprintf(w, `["%s.read_write"]`, gcpcredential.StorageScopePrefix)
+		} else if emailEndpoint == r.URL.Path {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprint(w, email)
+		} else if tokenEndpoint == r.URL.Path {
+			w.WriteHeader(http.StatusOK)
+			w.Header().Set("Content-Type", "application/json")
+			bytes, err := json.Marshal(token)
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			fmt.Fprintln(w, string(bytes))
+		} else if serviceAccountsEndpoint == r.URL.Path {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, "default/\ncustom")
+		} else {
+			http.Error(w, "", http.StatusNotFound)
+		}
+	}))
+	defer server.Close()
+	// Make a transport that reroutes all traffic to the example server
+	transport := utilnet.SetTransportDefaults(&http.Transport{
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			return url.Parse(server.URL + req.URL.Path)
+		},
+	})
+	provider := MakeRegistryProvider(transport)
+	response, err := GetResponse(provider)
+	if err != nil {
+		t.Errorf("Unexpected error while getting response: %s", err.Error())
+	}
+	if hasURL(registryURL, response) == false {
+		t.Errorf("URL %s expected in response, not found (response: %s)", registryURL, response.Auth)
+	}
+	for _, auth := range response.Auth {
+		if usernameMatches(expectedUsername, auth) == false {
+			t.Errorf("Expected username %s not found (username: %s)", expectedUsername, auth.Username)
+		}
+		if passwordMatches(dummyToken, auth) == false {
+			t.Errorf("Expected password %s not found (password: %s)", dummyToken, auth.Password)
+		}
+	}
+}
+
+func TestConfigProvider(t *testing.T) {
+	// Taken from from pkg/credentialprovider/gcp/metadata_test.go in k/k
+	registryURL := "hello.kubernetes.io"
+	email := "foo@bar.baz"
+	username := "foo"
+	password := "bar" // Fake value for testing.
+	auth := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%s:%s", username, password)))
+	sampleDockerConfig := fmt.Sprintf(`{
+   "https://%s": {
+     "email": %q,
+     "auth": %q
+   }
+}`, registryURL, email, auth)
+
+	const probeEndpoint = "/computeMetadata/v1/"
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
As a followup PR, maybe introduce a "fake metadata server" that can be used by all tests?

Reply:
Once the metadata server is decoupled from these tests, I believe it should be trivial to consolidate the three tests into a single table test. Right now there is a lot of copy-paste between the tests which could be eliminated. It would also become easier to add more tests cases.

Reply:
I think that would be good, definitely - I know some tests in the main repo would benefit from this as well.